tracer/src/Datadog.Trace/Ci/IntelligentTestRunnerClient.cs

```diff
@@ -680,6 +593,55 @@ private async Task<ObjectPackFilesResult> GetObjectsPackFileFromWorkingDirectory
         return new ObjectPackFilesResult(lstFiles.ToArray(), temporaryFolder);
     }
 
+    private void SetRequestHeader(IApiRequest request, bool useApplicationHeader)
+    {
+        request.AddHeader(HttpHeaderNames.TraceId, _id);
+        request.AddHeader(HttpHeaderNames.ParentId, _id);
+        if (_useEvpProxy)
+        {
+            request.AddHeader(EvpSubdomainHeader, "api");
+            if (useApplicationHeader)
+            {
+                request.AddHeader(EvpNeedsApplicationKeyHeader, "true");
+            }
+        }
+        else
+        {
+            request.AddHeader(ApiKeyHeader, _settings.ApiKey);
+            if (useApplicationHeader)
+            {
+                request.AddHeader(ApplicationKeyHeader, _settings.ApplicationKey);
+            }
+        }
+    }
+
+    private void CheckResponseStatusCode(IApiResponse response, string responseContent, bool finalTry)
+    {
+        // Check if the rate limit header was received.
+        if (response.StatusCode == 429 &&
+            response.GetHeader("x-ratelimit-reset") is { } strRateLimitDurationInSeconds &&
+            int.TryParse(strRateLimitDurationInSeconds, out var rateLimitDurationInSeconds))
+        {
+            if (rateLimitDurationInSeconds > 30)
+            {
+                // If 'x-ratelimit-reset' is > 30 seconds we cancel the request.
+                throw new RateLimitException();
+            }
+
+            throw new RateLimitException(rateLimitDurationInSeconds);
+        }
+
+        if (response.StatusCode is < 200 or >= 300 && response.StatusCode != 404)
+        {
+            if (finalTry)
+            {
+                Log.Error<int, string>("Request failed with status code {StatusCode} and message: {ResponseContent}", response.StatusCode, responseContent);
+            }
```

Comment:
You will no longer be logging if you get a rate limit exception when `finalTry=true`. Probably not a big deal, as long as you log in the calling method?

Reply:
I think we're already logging in the WithRetries method with the message "An error occurred while sending intelligent test runner data after {Retries} retries." so maybe it's okay that we don't log here